packages/tools/viewer/src/viewer.ts

```diff
@@ -668,32 +714,57 @@ export class Viewer implements IDisposable {
     }
 
     private get _activeAnimation(): Nullable<AnimationGroup> {
-        return this._details.model?.animationGroups[this._selectedAnimation] ?? null;
+        return this._modelInfo?.assetContainer.animationGroups[this._selectedAnimation] ?? null;
     }
 
     /**
      * The list of material variant names for the currently loaded model.
      */
     public get materialVariants(): readonly string[] {
-        return this._materialVariantsController?.variants ?? [];
+        return this._modelInfo?.materialVariantsController?.variants ?? [];
     }
 
     /**
      * The currently selected material variant.
      */
     public get selectedMaterialVariant(): Nullable<string> {
-        return this._materialVariantsController?.selectedVariant ?? null;
+        return this._modelInfo?.materialVariantsController?.selectedVariant ?? null;
     }
 
     public set selectedMaterialVariant(value: string) {
-        if (value !== this.selectedMaterialVariant && this._materialVariantsController?.variants.includes(value)) {
+        if (value !== this.selectedMaterialVariant && this._modelInfo?.materialVariantsController?.variants.includes(value)) {
             this._snapshotHelper.disableSnapshotRendering();
-            this._materialVariantsController.selectedVariant = value;
+            this._modelInfo.materialVariantsController.selectedVariant = value;
             this._snapshotHelper.enableSnapshotRendering();
             this.onSelectedMaterialVariantChanged.notifyObservers();
         }
     }
 
+    protected _beginLoadOperation(): IDisposable & { progress: Nullable<number> } {
+        // eslint-disable-next-line @typescript-eslint/no-this-alias
+        const viewer = this;

```

Comment:
I assume this is needed just because of the setter? I don't have any fundamental problem with it, but there is a way to define an arrow-function-setter using Object.defineProperty if that makes more sense:

```javascript
const loadOperation = {
            dispose: () => {
                this._loadOperations.delete(loadOperation);
                this.onLoadingProgressChanged.notifyObservers();
            },
        };
Object.defineProperty(loadOperation, "progress", {
  get: () => progress,
  set: (value: Nullable<number>) => {
              progress = value;
              this.onLoadingProgressChanged.notifyObservers();
          }
});
```

Reply:
Yes, it is because of the setter. Good to know `defineProperty` is an option, but I don't see any way to do this in a typesafe way. From a type standpoint, typescript does not know the `loadOperation` object has a `progress` property after `defineProperty` is called. I think I'll leave it the way I have it for better type safety.